### PR TITLE
ENG-988 improve GraphQL error handling

### PR DESCRIPTION
**Pull request by zreigz**
## Summary
The GraphQL client can return errors with different formats (regular string or JSON object). The new method `GetErrorResponse(err error, methodName string)` unifies the client error message.
**Old**
```
$ plural shell purge

2022/11/15 13:28:30 {"networkErrors":null,"graphqlErrors":[{"message":"no shell available for this user","path":["deleteShell"],"locations":[{"line":2,"column":2}]}]}
```
**New**
```
$ plural shell purge

2022/11/15 16:21:20 GraphQL error: no shell available for this user: DeleteShell
```
Network error:

```
2022/11/15 16:25:44 Network error: Response body {"message":"invalid_token"}: ListArtifacts
```

Very often the GraphQL server returns an` unknown error`
TODO:
Improve error messages on the API side:

```
$ plural api list artifacts aa
2022/11/15 16:22:44 GraphQL error: unknown error: ListArtifacts

```

## Checklist

- [ ] If required, I have updated the Plural documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a meaningful title and summary to convey the impact of this PR to a user.
- [ ] I have added relevant labels to this PR to help with categorization for release notes.

### CODE REVIEW

File: pkg/bundle/oidc.go [hunk 1]
```diff
@@ -49,8 +49,8 @@ func configureOidc(repo string, client api.Client, recipe *api.Recipe, ctx map[s
 		},
 	}
 	mergeOidcAttributes(inst, oidcSettings)
-
-	return client.OIDCProvider(inst.Id, oidcSettings)
+	err = client.OIDCProvider(inst.Id, oidcSettings)
+	return api.GetErrorResponse(err, "OIDCProvider")
```
Comment (DavidSpek) [inline pkg/bundle/oidc.go:53]: Shouldn't we check for `if err != nil` here. Similar to in [pkg/crypto/age.go](https://github.com/pluralsh/plural-cli/pull/277/files#diff-42e3a757ec072a799849e121c4cd26da264e442216ec0369c1f7c477fdd983f9)

File: cmd/plural/shell.go [hunk 2]
```diff
@@ -77,5 +77,6 @@ func handleShellPurge(c *cli.Context) error {
 	}
 
 	client := api.NewClient()
-	return client.DeleteShell()
+	err := client.DeleteShell()
+	return api.GetErrorResponse(err, "DeleteShell")
```
Comment (DavidSpek) [inline cmd/plural/shell.go:81]: Same here. Should we not check if the error is nil first, just like in [pkg/crypto/age.go](https://github.com/pluralsh/plural-cli/pull/277/files#diff-42e3a757ec072a799849e121c4cd26da264e442216ec0369c1f7c477fdd983f9).

File: cmd/plural/repos.go [hunk 3]
```diff
@@ -45,14 +45,15 @@ func (p *Plural) reposCommands() []cli.Command {
 
 func (p *Plural) handleUnlockRepo(c *cli.Context) error {
 	p.InitPluralClient()
-	return p.UnlockRepository(c.Args().First())
+	err := p.UnlockRepository(c.Args().First())
+	return api.GetErrorResponse(err, "UnlockRepository")
```
Comment (DavidSpek) [inline cmd/plural/repos.go:49]: Here as well, is a nil check not needed?

File: cmd/plural/push.go [hunk 4]
```diff
@@ -208,13 +208,14 @@ func (p *Plural) handleArtifact(c *cli.Context) error {
...
 }
 
 func (p *Plural) createCrd(c *cli.Context) error {
 	p.InitPluralClient()
 	fullPath, _ := filepath.Abs(c.Args().Get(0))
 	repo := c.Args().Get(1)
 	chart := c.Args().Get(2)
-	return p.CreateCrd(repo, chart, fullPath)
+	err := p.CreateCrd(repo, chart, fullPath)
+	return api.GetErrorResponse(err, "CreateCrd")
```
Comment (DavidSpek) [inline cmd/plural/push.go:220]: Here as well, is a nil check not needed?

File: cmd/plural/push.go [hunk 5]
```diff
@@ -208,13 +208,14 @@ func (p *Plural) handleArtifact(c *cli.Context) error {
 	input.Platform = c.String("platform")
 	input.Arch = c.String("arch")
 	_, err = p.CreateArtifact(c.Args().Get(1), input)
-	return err
+	return api.GetErrorResponse(err, "CreateArtifact")
```
Comment (DavidSpek) [inline cmd/plural/push.go:211]: Here as well, is a nil check not needed?

File: cmd/plural/push.go [hunk 6]
```diff
@@ -190,7 +190,7 @@ func (p *Plural) handleRecipeUpload(c *cli.Context) error {
 	}
 
 	_, err = p.CreateRecipe(c.Args().Get(1), recipeInput)
-	return err
+	return api.GetErrorResponse(err, "CreateRecipe")
```
Comment (DavidSpek) [inline cmd/plural/push.go:193]: Here as well, is a nil check not needed?

File: cmd/plural/packages.go [hunk 7]
```diff
@@ -61,15 +63,15 @@ func (p *Plural) uninstallPackage(c *cli.Context) error {
...
+				return api.GetErrorResponse(p.Client.UninstallTerraform(inst.Id), "UninstallTerraform")
 			}
 		}
 	}
 
 	if t == "helm" {
 		for _, inst := range space.Charts {
 			if inst.Chart.Name == name {
-				return p.Client.UninstallChart(inst.Id)
+				return api.GetErrorResponse(p.Client.UninstallChart(inst.Id), "UninstallChart")
```
Comment (DavidSpek) [inline cmd/plural/packages.go:74]: Here as well, is a nil check not needed?

File: cmd/plural/packages.go [hunk 8]
```diff
@@ -61,15 +63,15 @@ func (p *Plural) uninstallPackage(c *cli.Context) error {
 	if t == "terraform" {
 		for _, inst := range space.Terraform {
 			if inst.Terraform.Name == name {
-				return p.Client.UninstallTerraform(inst.Id)
+				return api.GetErrorResponse(p.Client.UninstallTerraform(inst.Id), "UninstallTerraform")
```
Comment (DavidSpek) [inline cmd/plural/packages.go:66]: Here as well, is a nil check not needed?

File: cmd/plural/api.go [hunk 9]
```diff
@@ -187,5 +186,6 @@ func (p *Plural) handleArtifacts(c *cli.Context) error {
 
 func (p *Plural) handleCreateDomain(c *cli.Context) error {
 	p.InitPluralClient()
-	return p.CreateDomain(c.Args().First())
+	err := p.CreateDomain(c.Args().First())
+	return api.GetErrorResponse(err, "CreateDomain")
```
Comment (DavidSpek) [inline cmd/plural/api.go:190]: Here as well, is a nil check not needed?

**Comment by DavidSpek** [comment]
Overall looks good. Just some small section where an `if err != nil` isn't being done. Not sure if it is necessary but just an inconsistency I noticed. Otherwise it looks good.

**Comment by michaeljguarino** [comment]
to handle the err == nil cases, if we just passthrough nil in the GraphQlError wrapper, it should be sufficient and prevent inelegant branching in the code

**Comment by zreigz** [comment]
> Overall looks good. Just some small section where an `if err != nil` isn't being done. Not sure if it is necessary but just an inconsistency I noticed. Otherwise it looks good.

generally, the method 'GetErrorResponse' digests the error and checks if it's nil. If not then converts to the new format

File: pkg/api/client.go [hunk 10]
```diff
@@ -97,3 +100,28 @@ func FromConfig(conf *config.Config) Client {
...
+	}
+	errResponse := &rawclient.ErrorResponse{}
+	newErr := json.Unmarshal([]byte(err.Error()), errResponse)
+	if newErr != nil {
+		return err
+	}
+
+	errList := errors.New(methodName)
+	if errResponse.GqlErrors != nil {
+		for _, err := range *errResponse.GqlErrors {
```
Comment (michaeljguarino) [inline pkg/api/client.go:116]: We should make a reduce function in pluralsh/polly to eliminate these common handwritten loops

**Pull request merged**